addons/account/models/account_move_line.py

```diff
@@ -1192,28 +1193,27 @@ def _compute_reconciled_lines_excluding_exchange_diff_ids(self):
             )
             line.reconciled_lines_excluding_exchange_diff_ids = all_lines - excluded_ids
 
-    @api.depends('sequence', 'move_id')
+    @api.depends('move_id.invoice_line_ids.sequence', 'move_id')
     def _compute_parent_id(self):
-        for _move, lines in self.grouped('move_id').items():
-            last_section = False
-            last_sub = False
-            for line in lines.sorted('sequence'):
-                if line.display_type == 'line_section':
-                    last_section = line
-                    line.parent_id = False
-                    last_sub = False
-                elif line.display_type == 'line_subsection':
-                    line.parent_id = last_section
-                    last_sub = line
-                elif line.display_type == 'product':
-                    line.parent_id = last_sub or last_section or False
-                else:
-                    line.parent_id = False
+        section_lines_ordered = (self.move_id.invoice_line_ids
+            .filtered(lambda line: line.display_type in ('line_section', 'line_subsection'))
+            .sorted('sequence')
+            .grouped('sequence'))
+        section_lines_ordered_sequences = list(section_lines_ordered)
+
+        for line in self:
+            if line.display_type == 'line_subsection':
+                line_index = bisect_left(section_lines_ordered_sequences, line.sequence, key=lambda line: section_lines_ordered[line][0].display_type == 'line_section')
```

Comment:
I can see that a subsection should only have a section as a parent, but maybe we should include a comment describing why this key works